src/rust/grapl-utils/src/rusoto_ext/dynamodb.rs

```diff
@@ -125,55 +121,175 @@ pub trait GraplDynamoDbClientExt: DynamoDb + Send + Sync {
+                response_map.into_iter().for_each(|(table_name, rows)| {
+                    total_responses
+                        .entry(table_name.clone())
+                        .or_insert(Vec::new())
+                        .extend(rows);
+                });
             }
 
             // if a response was too large, we may have unprocessed keys.
             // we can just directly insert these back into pending items for processing
-            match response.unprocessed_keys {
-                Some(unprocessed_keys) => {
-                    let unprocessed_keys: Vec<_> = unprocessed_keys
-                        .into_iter()
-                        .flat_map(|(table_name, keys_and_attributes)| {
-                            keys_and_attributes
-                                .keys
-                                .into_iter()
-                                .map(|row_keys| (table_name.clone(), row_keys))
-                                .collect::<Vec<_>>()
-                        })
-                        .collect();
+            if let Some(unprocessed_keys) = response.unprocessed_keys {
+                let unprocessed_keys: Vec<_> = unprocessed_keys
+                    .into_iter()
+                    .flat_map(|(table_name, keys_and_attributes)| {
+                        keys_and_attributes
+                            .keys
+                            .into_iter()
+                            .map(|row_keys| (table_name.clone(), row_keys))
+                            .collect::<Vec<_>>()
+                    })
+                    .collect();
 
-                    pending_items.extend(unprocessed_keys);
-                }
-                _ => {}
+                pending_items.extend(unprocessed_keys);
             }
         }
 
-        Ok(BatchGetItemOutput {
+        let output = BatchGetItemOutput {
             consumed_capacity: batch_get_items_input
                 .return_consumed_capacity
                 .map(|_| total_consumed_capacity),
             responses: Some(total_responses),
             unprocessed_keys: None,
-        })
+        };
+
+        Ok(output)
+    }
+
+    /// Similar to [`batch_get_item_reliably`], this function handles chunking requests to avoid going
+    /// over limits imposed on the size of BatchWriteItem operations in DynamoDB.
+    async fn batch_write_item_reliably(
+        &self,
+        batch_write_item: BatchWriteItemInput
+    ) -> Result<BatchWriteItemOutput, RusotoError<BatchWriteItemError>> {
+        let BatchWriteItemInput {
+            request_items,
+            return_consumed_capacity,
+            return_item_collection_metrics
+        } = batch_write_item;
+
+        let mut pending_writes: Vec<_> = request_items.into_iter()
+            .flat_map(|(table_name, write_requests)| {
+                write_requests.into_iter()
+                    .map(|write_request| (table_name.clone(), write_request))
+                    .collect::<Vec<_>>()
```

Comment:
I think '.collect' is unnecessary. `flat_map` takes IntoIterator, so we can avoid allocating to a Vec just to turn it back into an iterator.